spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -790,4 +804,53 @@ public Identifier identifier() {
   public static TableIdentifier identifierToTableIdentifier(Identifier identifier) {
     return TableIdentifier.of(Namespace.of(identifier.namespace()), identifier.name());
   }
+
+  /**
+   * Use Spark to list all partitions in the table.
+   *
+   * @param spark a Spark session
+   * @param rootPath a table identifier
+   * @param format format of the file
+   * @return all table's partitions
+   */
+  public static List<SparkTableUtil.SparkPartition> getPartitions(SparkSession spark, Path rootPath, String format) {
+    FileStatusCache fileStatusCache = FileStatusCache.getOrCreate(spark);
+    Map<String, String> emptyMap = Collections.emptyMap();
+
+    InMemoryFileIndex fileIndex = new InMemoryFileIndex(
```

Comment:
We can't use the version in SparkUtils because unfortunately the Scala api for this method has added additional default args. In Scala this wouldn't be a problem, but in java we need to explicitly call the function with all args.

Reply:
Could we use reflection for this? Seems we had a similar issue in `HiveClientPool`.

I understand `InMemoryFileIndex` is a low-level API and it may not be the best idea to rely on it. I'm ok with this as the first step. We may build our own logic if we need it in other query engines.

I'd be also alright to always require a specific partition to import and not support importing all partitions in a given path, which should avoid the need to use Spark's `InMemoryFileIndex` to infer the underlying partitioning. As I wrote before, I think the primary use case for this is CREATE TABLE LIKE and then calling `add_files` for a couple of partitions to try reads/writes or migrating partition by partition if a table has too many partitions to migrate them at once. In both cases, there is a specific partition we have to import.

If there is a need to import files in a given location, one can create an external Spark table pointing to the location, call `MSCK REPAIR TABLE` to infer partitions and add them to the metastore and then call either snapshot or migrate procedure.

What are your thoughts, @RussellSpitzer @rdblue?

Reply:
One scenario where we may need to infer partitioning is `migrate('path/to/hadoop/table')`.

Reply:
I'd prefer not to infer partitioning. I left a comment on #2068. The Hive import works because we know that the partitions match. Inferring too much seems like it would lead to bad data to me.

Reply:
I think adding Iceberg files through this API is off the table. The only case we are talking about is the import of existing non-Iceberg tables/partitions. In this PR, Russell uses `InMemoryFileIndex` to get a list of partitions for a given root table location.

The question I was debating in my previous comment is whether we need to support importing root table locations. As I said earlier, the main use case for `add_files` will probably be to add a few partitions after CREATE TABLE LIKE if the original non-Iceberg table contains a lot of partitions and calling SNAPSHOT may be expensive. Another use case is gradual migration. In that case, we don't need to infer what partitions are present. We can have the following procedure:

```
iceberg.system.add_files(
  table => 'db.tbl', -- required
  path => '/root/table/location', -- required
  format => 'parquet', -- required
  partition => map('p1', 'v1', 'p2', 'v2') -- required for partitioned tables
)
```

In this case, we just combine the partition tuple with the root location and import a single partition at a time.

The question is whether we should make `partition` optional for partitioned tables so that you can just point to the root table location. If yes, we will have to use something like `InMemoryFileIndex` or our own solution to find out which partitions are present in that location. I would like to avoid that but we need to come up with a solution for importing non-HMS tables.

For example, I think this should be valid:

```
MIGRATE `parquet`.`/path/to/table/on/hdfs`
USING iceberg
```

It seems we will need to know what partitions are present in that case.

Reply:
I agree that migrating a location-based table is a valid use case. To me, that's slightly different because like Hive it is a "defined" table format. We would also convert the detected partition structure to a partition spec that uses identity, so it is predictable and limited.

For importing files, I don't want to get into a situation where we have to support non-identity partition transforms or guess how to match up the partitioning. I think the safest thing is to import a single partition at a time, but that would be a pain. Maybe if we can come up with reasonable rules for this we can make it work:
1. All partitions must be identity partitions
2. No transforming data except to parse numbers

Others?

Reply:
Alright, I think we all agree that `add_files` procedure should be limited to identity partitioning and, ideally, it should not infer partitions (i.e. we should not try to build a list of partitions in a given location). The primary use case for this procedure is partial migration and people can use SNAPSHOT or MIGRATE if they need to migrate all partitions at the same time. Instead of inferring a list of partitions ourselves, we better ask the user which partitions to import.

That means we don't need to use `InMemoryFileIndex` right now. We will come back to it while adding support for migrating path-based tables. We should also take other query engines into account.

The open question right now is whether we should support adding multiple partitions in the same call? That seems beneficial but how do we do this and what would it mean for the procedure API? In earlier discussions, @rdblue mentioned supporting something like `part_col1=*/part_col2=x` would be great. While it would be easier for us to support only a single partition, I'd vote for a better user experience.

We could represent partition path as a string with wildcards but I'd suggest using a map.

```
iceberg.system.add_files(
  table => 'db.tbl', -- required
  path => '/root/table/location', -- required
  format => 'parquet', -- required
  partition => map('p1', '*', 'p2', 'v2') -- required for partitioned tables
)
```

We may add support for `*` later but using a map instead of a string does not require us parsing the path to get individual partitions.

Reply:
Thoughts, @rdsr @RussellSpitzer @rdblue?

Reply:
@aokolnychyi is the order among the partition columns assumed? I slightly prefer @rdblue's where order seems explicit

Reply:
@rdsr, I think the procedure can load the Iceberg table and use its spec to determine the correct order of partition columns. That seems like the most reliable way and does not require parsing partition strings.

@pvary brought up a good point about custom partition locations in [this](https://github.com/apache/iceberg/pull/2210#discussion_r578317436) comment. 

The API I proposed in the comment above won't work for that. @RussellSpitzer and I had a quick chat and it seems we can support the following:

```
iceberg.system.add_files(
  source_table => 'source_table' -- required
  table => 'db.tbl', -- required
  partition => map('p1', '*', 'p2', 'v2') -- optional
)
```

The procedure can be called like this:

```
iceberg.system.add_files(
  source_table => `parquet`.`path/to/table`
  table => 'iceberg.db.tbl',
  partition => map('p1', '*', 'p2', 'v2')
)
```

```
iceberg.system.add_files(
  source_table => `db.hive_tbl`
  table => 'db.iceberg_tbl',
  partition => map('p1', '*', 'p2', 'v2')
)
```

This way, we should support custom partition locations as we will go to HMS for Hive tables.

Reply:
Sounds good @aokolnychyi !

Reply:
For the path-based option, would that load the path-based table using Spark and then get the files from it? If so, I like delegating the partition inference like that. I guess my question is: what if `path/to/table` isn't a partition directory and is further up the file tree?

Reply:
@rdblue, I think `parquet`.`path/to/table` would need to point to the root table location all the time, just like in SQL on file in Spark. The `partition` map would tell us which partitions to import.

Let's consider the following call:

```
iceberg.system.add_files(
  source_table => `parquet`.`path/to/table`
  table => 'iceberg.db.tbl',
  partition => map('p1', 'v1', 'p2', 'v2')
)
```

I think the procedure would detect the identifier points to a path-based Parquet table and would take the root table location as a basis and add a path for the partition. In other words, the call above would be translated into importing a single Parquet partition.

```
path/to/table/p1=v1/p2=v2
```

Does that make sense?

Reply:
For HMS tables, we will need to go to the metastore and find the location of `p1=v1, p2=v2` partition.

Reply:
Supporting bulk imports can be done by expanding `*` in the partition values.

For example, `map('p1', '*', 'p2', 'v2')` would import all partitions where `p2=v2`.

Reply:
I think that Spark allows pointing to a location that is inside a path-based table. So if you pointed to p1=x, then you'd get p2 and p3 columns, but not p1. We can test that, but that's what I'm concerned about. If we need a value for p1, but don't get it from Spark then there is no way to add it besides including it in the partition map.

Maybe using the map as a filter like you suggest would work, but that seems a bit strange to me because it is sometimes supplying partition values and sometimes getting those values from the path.

Reply:
I've switched this around and now we always determine partitions via filtering, we never explicitly add the path but do rely on the Spark methodology to determine which partitions exist. This means for Hive tables we ask hive, and for file based tables we use the in memory file index.

if you specify a sub directory as your location i'll have to test what our current code does, i'm not sure what the outcome would be

Reply:
I thought we would not need to use Spark's in-memory file index for path-based tables to determine which partitions are inside if we use this approach. The main reason is that path-based tables cannot have custom table locations and we can always construct a partition path if we have a partition map and the root table location. Also, using the in-memory file index is expensive as it will build us a map of all partitions even thought we neew a couple of them and we can construct partition paths ourselves.

@rdblue, I think your concerns can be addressed by extra validation. We will have access to the Iceberg partition spec. We can validate the map of partition values contains all columns.

Reply:
I am open to other alternatives but this idea seems to be the easiest among the options we discussed and we can actually avoid using the in-memory file index for now.

Reply:
I think in-memroy file index is fine for the moment, but we can always change it later, now that it's an internal part of just the file based approach it should be easy to swap out.

Reply:
I had offline discussions with @RussellSpitzer and @rdblue and there is no ideal solution. Probably, we won't get around inferring partitions and since we will need that in other places, so we can probably support it now.

It is fine to start with Spark's `InMemoryFileIndex` for now. However, we should migrate to our own solution and leverage it in other query engines during snapshot and migrate commands. I've created #2300 for this.

Reply:
I am going to resolve this thread to simplify the review. Feel free to reopen as needed.

Reply:
Thanks everybody for the discussion!